### PR TITLE
update script url from builder image labels with s2i build --as-dockerfile

### PR DESCRIPTION
**Pull request by gabemontero**
see https://github.com/shipwright-io/build/issues/810

### CODE REVIEW

**Comment by coreydaley** [comment]
Is there a test that should be added or updated to verify that we don't have a regression?

**Comment by gabemontero** [comment]
> Is there a test that should be added or updated to verify that we don't have a regression?

that entire package is absent of tests ... 7 files in all

I'm citing lack of cycles for taking that on as part of this PR, at least for the next week or so

and when you factor in the age of this stuff ... but yes, general convention, recent aos-devel emails, etc., we want more unit test coverage 

so if between you and @adambkaplan and myself we insist on addressing that now with this PR, either @jfdenise will have to wait a bit longer until I have cycles for that, or we collectively collaborate with the three of us from build-api contributing to this PR to accelerate things

**Comment by gabemontero** [comment]
Pending that, I think we've got an instance of our old jenkins CI based tests using a very old PHP image:

```
I0623 20:50:05.157523   16454 sti.go:711] Parse error: syntax error, unexpected '?', expecting variable (T_VARIABLE) in /opt/app-root/src/vendor/cakephp/plugin-installer/src/Plugin.php on line 157
I0623 20:50:05.157523   16454 sti.go:715] PHP Parse error:  syntax error, unexpected '?', expecting variable (T_VARIABLE) in /opt/app-root/src/vendor/cakephp/plugin-installer/src/Plugin.php on line 157
I0623 20:50:05.228099   16454 docker.go:1044] Waiting for container "9af174f2fb74b76e7dcb87602a48ed63370cb2afc46cd572fbbe9aeb31a97f69" to stop ...
I0623 20:50:05.581118   16454 docker.go:980] Removing container "9af174f2fb74b76e7dcb87602a48ed63370cb2afc46cd572fbbe9aeb31a97f69" ...
I0623 20:50:05.617967   16454 docker.go:990] Removed container "9af174f2fb74b76e7dcb87602a48ed63370cb2afc46cd572fbbe9aeb31a97f69"
I0623 20:50:05.618464   16454 cleanup.go:33] Removing temporary directory /tmp/s2i203148722
I0623 20:50:05.618485   16454 fs.go:313] Removing directory '/tmp/s2i203148722'
I0623 20:50:05.625525   16454 build.go:172] Build failed
E0623 20:50:05.625550   16454 errors.go:296] An error occurred: non-zero (13) exit code from docker.io/centos/php-70-centos7

```

the oldest version we ship in OCP is 7.2 .... 7.0 EOL'ed and was removed back in the OCP 4.4 timeframe !!

**Comment by gabemontero** [comment]
I'll see about fixing that

**Comment by gabemontero** [comment]
ok fingers crossed

**Comment by gabemontero** [comment]
New debug using @jfdenise 's builder image ... you see `S2I Scripts URL:		image:///usr/local/s2i`:

```
I0623 16:32:06.936601  397602 build.go:167] 
Source:				https://github.com/jfdenise/wildfly-s2i
Context Directory:		test/test-app
Output Image Tag:		
Environment:			
Labels:				
Incremental Build:		disabled
Remove Old Build:		disabled
Builder Pull Policy:		if-not-present
Previous Image Pull Policy:	if-not-present
Quiet:				disabled
Layered Build:			disabled
Artifacts Destination:		/tmp
S2I Scripts URL:		image:///usr/local/s2i
Docker Endpoint:		unix:///var/run/docker.sock
Docker Pull Config:		/home/gmontero/.docker/config.json
Docker Pull User:		gabemontero

```

and 

```
I0623 16:32:07.535764  397602 install.go:261] Using "assemble" installed from "image:///usr/local/s2i/assemble"
I0623 16:32:07.535796  397602 install.go:261] Using "run" installed from "image:///usr/local/s2i/run"
I0623 16:32:07.535811  397602 install.go:261] Using "save-artifacts" installed from "image:///usr/local/s2i/save-artifacts"

```

**Comment by gabemontero** [comment]
hmmm ... the prior errors I noted in https://github.com/openshift/source-to-image/pull/1068#issuecomment-867157392 went away, and the test log seems to convey everything worked, but then it just says it failed at the end:

```
########## STARTING STAGE: EXECUTE S2I RUNIMAGE TESTS ##########
+ [[ -s /var/lib/jenkins/jobs/test_pull_request_s2i_master/workspace/activate ]]
+ source /var/lib/jenkins/jobs/test_pull_request_s2i_master/workspace/activate
++ export VIRTUAL_ENV=/var/lib/jenkins/origin-ci-tool/95a472392910104fc805c4ab8af39791105c956d
++ VIRTUAL_ENV=/var/lib/jenkins/origin-ci-tool/95a472392910104fc805c4ab8af39791105c956d
++ export PATH=/var/lib/jenkins/origin-ci-tool/95a472392910104fc805c4ab8af39791105c956d/bin:/sbin:/usr/sbin:/bin:/usr/bin
++ PATH=/var/lib/jenkins/origin-ci-tool/95a472392910104fc805c4ab8af39791105c956d/bin:/sbin:/usr/sbin:/bin:/usr/bin
++ unset PYTHON_HOME
++ export OCT_CONFIG_HOME=/var/lib/jenkins/jobs/test_pull_request_s2i_master/workspace/.config
++ OCT_CONFIG_HOME=/var/lib/jenkins/jobs/test_pull_request_s2i_master/workspace/.config
++ mktemp
+ script=/tmp/tmp.KMCMNr8BQp
+ cat
+ chmod +x /tmp/tmp.KMCMNr8BQp
+ scp -F /var/lib/jenkins/jobs/test_pull_request_s2i_master/workspace/.config/origin-ci-tool/inventory/.ssh_config /tmp/tmp.KMCMNr8BQp openshiftdevel:/tmp/tmp.KMCMNr8BQp
+ ssh -F /var/lib/jenkins/jobs/test_pull_request_s2i_master/workspace/.config/origin-ci-tool/inventory/.ssh_config -t openshiftdevel 'bash -l -c "timeout 7200 /tmp/tmp.KMCMNr8BQp"'
+ cd /data/src/github.com/openshift/source-to-image
+ hack/test-stirunimage.sh
working dir:  /tmp/sti/test-work.z2Ii
s2i working dir:  /tmp/sti/test-work.z2Ii
/tmp/sti/test-work.z2Ii /data/src/github.com/openshift/source-to-image

cloning source into working dir


TEST PASSED


s2i build with relative path without file://


TEST PASSED


s2i build with volume options


TEST PASSED

/data/src/github.com/openshift/source-to-image

s2i build with absolute path with file://


TEST PASSED


s2i build with absolute path without file://


TEST PASSED


s2i build with non-git repo file location


TEST PASSED

I0623 21:16:16.332697   18102 download.go:38] Copying sources from "/tmp/sti/test-work.z2Ii/cakephp-ex" to "/tmp/s2i659742621/upload/src"

TEST PASSED


s2i rebuild


TEST PASSED


TEST PASSED


s2i usage


TEST PASSED

I0623 21:17:14.613868   19524 sti.go:711] Sample invocation:

TEST PASSED


s2i build with overriding assemble/run scripts

Dumping logs since did not run successfully before cleanup of /tmp/sti/test-work.z2Ii ...
---> Installing application source...
=> sourcing 20-copy-config.sh ...
---> 21:17:19     Processing additional arbitrary httpd configuration provided by s2i ...
=> sourcing 00-documentroot.conf ...
=> sourcing 50-mpm-tuning.conf ...
=> sourcing 40-ssl-certs.sh ...
Build completed successfully

Complete
++ export status=FAILURE
++ status=FAILURE
+ set +o xtrace
########## FINISHED STAGE: FAILURE: EXECUTE S2I RUNIMAGE TESTS [00h 02m 03s] ##########
```

I'll pick back up tomorrow

File: pkg/cmd/cli/cmd/generate.go
Comment (adambkaplan) [inline pkg/cmd/cli/cmd/generate.go:99]: Create the context object when the root command is invoked, and pass it through here:

```suggestion
func manageConfigImageLabelsBuildImageName(ctx context.Context, cfg *api.Config) error {
```

File: pkg/cmd/cli/cmd/generate.go
Comment (adambkaplan) [inline pkg/cmd/cli/cmd/generate.go:108]: Use passed in context per recommendation above.

File: pkg/cmd/cli/cmd/build.go
Comment (adambkaplan) [inline pkg/cmd/cli/cmd/build.go:164]: Will need to create and pass in a context with my recommended refactoring

**Review by adambkaplan — CHANGES_REQUESTED** [review]

**Comment by adambkaplan** [comment]
Test failure seems to be this:

```
s2i build with overriding assemble/run scripts

Dumping logs since did not run successfully before cleanup of /tmp/sti/test-work.z2Ii ...
---> Installing application source...
=> sourcing 20-copy-config.sh ...
---> 21:17:19     Processing additional arbitrary httpd configuration provided by s2i ...
=> sourcing 00-documentroot.conf ...
=> sourcing 50-mpm-tuning.conf ...
=> sourcing 40-ssl-certs.sh ...
Build completed successfully
```

File: pkg/cmd/cli/cmd/build.go
Comment (gabemontero) [reply pkg/cmd/cli/cmd/build.go:164]: are you fine with creating a `context.Background()` above this and then passing it into `manageConfigImageLabelsBuildImageName` @adambkaplan ?

Comment (adambkaplan) [reply pkg/cmd/cli/cmd/build.go:164]: That would work

**Comment by gabemontero** [comment]
OK turns out `registry.access.redhat.com/ubi8/php-74:latest` has our s2i image labels on it ... from an inspect of the image, I won't paste the who thing, but 

```
                "io.s2i.scripts-url": "image:///usr/libexec/s2i",
```

is in the labels section.  FWIW the same holds true for the old image `docker.io/centos/php-70-centos7`.  So our change "fixing" the use of those labels on the builder image and changing the behavior makes sense.

But that also means that the builder image labels are taking precedence over the scripts found via --context-dir options.  It feels like any scripts found via `--context-dir` should take precedence over the builder image labels, though I'm not finding any explicit docs on such an order of precedence, vs. say the explicit --scripts-url option taking precedence.

I'll move forward on sorting out and fixing the precedence with --context-dir putting those ahead of builder image labels.

**Comment by gabemontero** [comment]
> OK turns out `registry.access.redhat.com/ubi8/php-74:latest` has our s2i image labels on it ... from an inspect of the image, I won't paste the who thing, but
> 
> ```
>                 "io.s2i.scripts-url": "image:///usr/libexec/s2i",
> ```
> 
> is in the labels section. FWIW the same holds true for the old image `docker.io/centos/php-70-centos7`. So our change "fixing" the use of those labels on the builder image and changing the behavior makes sense.
> 
> But that also means that the builder image labels are taking precedence over the scripts found via --context-dir options. It feels like any scripts found via `--context-dir` should take precedence over the builder image labels, though I'm not finding any explicit docs on such an order of precedence, vs. say the explicit --scripts-url option taking precedence.
> 
> I'll move forward on sorting out and fixing the precedence with --context-dir putting those ahead of builder image labels.

actually as I try things out I realized the presene of `--context-dir` does not really matter .... they could have a .s2i/bin off of the root of their source

it is really a question of whether the source .s2i/bin  should take precedence over the builder image label, or if we make them use --script-url  to completely override the builder label

time for an office hours / slack discussion

**Comment by gabemontero** [comment]
> > OK turns out `registry.access.redhat.com/ubi8/php-74:latest` has our s2i image labels on it ... from an inspect of the image, I won't paste the who thing, but
> > ```
> >                 "io.s2i.scripts-url": "image:///usr/libexec/s2i",
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > is in the labels section. FWIW the same holds true for the old image `docker.io/centos/php-70-centos7`. So our change "fixing" the use of those labels on the builder image and changing the behavior makes sense.
> > But that also means that the builder image labels are taking precedence over the scripts found via --context-dir options. It feels like any scripts found via `--context-dir` should take precedence over the builder image labels, though I'm not finding any explicit docs on such an order of precedence, vs. say the explicit --scripts-url option taking precedence.
> > I'll move forward on sorting out and fixing the precedence with --context-dir putting those ahead of builder image labels.
> 
> actually as I try things out I realized the presene of `--context-dir` does not really matter .... they could have a .s2i/bin off of the root of their source
> 
> it is really a question of whether the source .s2i/bin should take precedence over the builder image label, or if we make them use --script-url to completely override the builder label
> 
> time for an office hours / slack discussion

I do have a commit that allows for scripts from the source to override scripts noted by the labels pushed here, to a) facilitate the team discussion I just noted, and b) confirm it passes the previously failing test like it did for my non docker daemon equivalent local testing

**Comment by gabemontero** [comment]
ok @adambkaplan @coreydaley I have both gen dockerfile from @jfdenise 's scenario  and the classic s2i build form from the jenkins-s2i CI tests using the builder image labels 

In the end I believe the order of precedence stuff is now consistent with that is previously there.  We can discuss that either here or in office hours as needed.  It was in the end a matter of where to insert additional checks while installing scripts.

I'll clean a few things up with my local hack session after lunch and update the PR.

**Comment by gabemontero** [comment]
OK @adambkaplan @coreydaley PR update PTAL

I expect the jenkins-s2i test to pass as it is passing for me locally now that I have docker installed again

Fixing the multiple permutations lead to changing files with associated unit tests, so in the end, I believe I've added coverage for the relevant pieces of the change in some form or fashion

**Comment by gabemontero** [comment]
> OK @adambkaplan @coreydaley PR update PTAL
> 
> I expect the jenkins-s2i test to pass as it is passing for me locally now that I have docker installed again

yep all CI tests are passing
> 
> Fixing the multiple permutations lead to changing files with associated unit tests, so in the end, I believe I've added coverage for the relevant pieces of the change in some form or fashion

File: pkg/cmd/cli/cmd/generate.go
Comment (adambkaplan) [inline pkg/cmd/cli/cmd/generate.go:103]: This needs a better godoc - I think a) we can error if the builder image pull spec is invalid, and b) probably is not necessary to say that the methods we call won't return an error. If their signature includes an error, we should assume that it can return an error.

```suggestion
// manageConfigImageLabelsBuildImageName extracts the image labels from builder image in the provided config.
// Returns an error if the builder image name is invalid or there is an error extracting the image labels.
func manageConfigImageLabelsBuildImageName(ctx context.Context, cfg *api.Config) error {
```

File: pkg/util/labels.go
Comment (adambkaplan) [inline pkg/util/labels.go:64]: nit
```suggestion
// AdjustConfigWithImageLabels examines previously retrieved builder image labels and returns
// any specified update to the scripts URL and destination settings.  Note, we do
```

**Review by adambkaplan — COMMENTED** [review]
/approve

Only needs a few godoc fixes.

**Comment by gabemontero** [comment]
/assign @coreydaley 

I've pulled in Adam's 2 suggested commits and squashed.  With him on PTO now, can you take a look and either provide any additional suggestions you have, or lgtm?

thanks

File: pkg/cmd/cli/cmd/build.go
Comment (coreydaley) [inline pkg/cmd/cli/cmd/build.go:166]:
```suggestion 
			err = manageConfigImageLabelsBuildImageName(context.Background(), cfg)
```

Comment (coreydaley) [reply pkg/cmd/cli/cmd/build.go:166]: Makes it more clear that this `context` is not used anywhere else.

File: pkg/cmd/cli/cmd/generate.go
Comment (coreydaley) [inline pkg/cmd/cli/cmd/generate.go:116]:
```suggestion
	if cfg.BuilderImageLabels, err = getImageLabels(ctx, ref); err != nil {
		return err
	}
```

I don't think there is a reason to create an additional variable here, should be able to assign directly to the future recipient.

File: pkg/util/labels.go
Comment (coreydaley) [inline pkg/util/labels.go:73]:
```suggestion
	if cfg.BuilderImageLabels == nil {
		return "", ""
	}
	
	scriptsURL, destination := "", ""
```
No need to create the variables before checking to see if we are just going to return two empty strings.

**Review by coreydaley — APPROVED** [review]
A couple of small nits. If you want to squash your commits I can lgtm this.

**Comment by gabemontero** [comment]
pulled in and squahsed @coreydaley thanks

**Comment by coreydaley** [comment]
/lgtm
You should be able to re-apply the lgtm after you run gofmt and commit it.

**Comment by gabemontero** [comment]
> /lgtm
> You should be able to re-apply the lgtm after you run gofmt and commit it.

can I?

curious, let's see

/lgtm

**Comment by gabemontero** [comment]
> @gabemontero: you cannot LGTM your own PR.

nope need a re-lgtm please @coreydaley

**Comment by gabemontero** [comment]
hmm ... our flurry of activity may have borked travis ... if it is still that way after the holiday we'll either get contact test platform our argue green button given it passed so often previoiusly and all we did since last commit is gofmt

**Comment by gabemontero** [comment]
oh I think it finally started

**Pull request merged**
